core/libp2p/basic/message_read_writer.hpp

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_MESSAGE_READ_WRITER_HPP
+#define KAGOME_MESSAGE_READ_WRITER_HPP
+
+#include <memory>
+#include <vector>
+
+#include "common/buffer.hpp"
+#include "libp2p/basic/message_read_writer_error.hpp"
+#include "libp2p/basic/readwriter.hpp"
+#include "libp2p/basic/varint_reader.hpp"
+#include "libp2p/multi/uvarint.hpp"
+
+namespace libp2p::basic {
+  /**
+   * Allows to read and write messages of (\tparam Message) type
+   * @tparam Message - type of the message; for now, works with Protobuf
+   * messages only
+   */
+  template <typename Message>
+  class MessageReadWriter {
```

Comment:
also this class can accept connection and error handler in constructor and work as a `session` in asio (e.g. `public std::enable_shared_from_this<MessageReadWriter>`).

doing it this way would let you to write a method
```
MessageReadWriter rw(conn, errhandler);
rw.read([](Message msg){
  // handle msg read;  
});
```

Reply:
Also, I would add second template argument - Serde (serializer/deserializer), and defined compile time detection for protobuf messages, so it would become
```
template <typename Message, typename Serde>
void read(...){
 ... actual implementation ...
}

template <typename Message>
auto read(...) -> enable_if<... Message is Protobuf... > {
   read<Message, ProtobufSerde>(...); 
}
```